### PR TITLE
Change uri_to_fname to not convert non-file URIs

### PR DESCRIPTION
**Pull request by mfussenegger**
A URI with a scheme other than file doesn't have a local file path.

This is an alternative to https://github.com/neovim/neovim/pull/12264 and
relates to https://github.com/neovim/neovim/issues/12210

### CODE REVIEW

**Comment by tjdevries** [comment]
This seems like the better solution. Does this make it so you can handle the `jdt:` protocol now?

**Comment by mfussenegger** [comment]
>  Does this make it so you can handle the jdt: protocol now?

Yes.

File: test/functional/lua/uri_spec.lua [hunk 1]
```diff
@@ -112,6 +112,13 @@ describe('URI methods', function()
         eq('C:\\xy\\åäö\\ɧ\\汉语\\↥\\🤦\\🦄\\å\\بِيَّ.txt', exec_lua(test_case))
       end)
     end)
+    describe('decode non-file uri', function()
+      it('uri_to_fname returns non-file uri unchanged', function()
+        eq('jdt://content/%5C/', exec_lua [[
+          return vim.uri_to_fname('jdt://content/%5C/')
+        ]])
+      end)
+    end)
```
Comment (h-michael) [inline test/functional/lua/uri_spec.lua:121]: Would you add a test case that uri has not scheme?

**Comment by h-michael** [comment]
@mfussenegger Can Windows handle `jdt://...` ?

File: test/functional/lua/uri_spec.lua (hunk 1 above)
Comment (mfussenegger) [reply test/functional/lua/uri_spec.lua:122]: Yep, added one.

**Comment by mfussenegger** [comment]
> @mfussenegger Can Windows handle `jdt://...` ?
> I think that the plat form is irrelevant because it is not a file path but a uri.

Yes should be platform independent. What I'm trying to achieve is that neovim doesn't transform URIs that don't have the file scheme to support this:

```

  LSP Server
    |
    | Returns URI as part of a response
  Neovim
    |
    | Loads buffer with URI
    └ Calls BufReadCmd
               |
               | Requests contents for URI from LSP server
```

Currently between "Server returns URI" and "Calls BufReadCmd" the URIs are being decoded, which causes the LSP server to not recognize or understand the URI any longer.

File: runtime/lua/vim/uri.lua [hunk 2]
```diff
@@ -76,6 +76,10 @@ local function uri_from_bufnr(bufnr)
 end
 
 local function uri_to_fname(uri)
+  local scheme = assert(uri:match("^([a-z]+)://.*"), 'Uri must contain a scheme: ' .. uri)
```
Comment (h-michael) [inline runtime/lua/vim/uri.lua:79]: nit
```suggestion
  local scheme = assert(uri:match("^([a-z]+)://.*"), 'URI must contain a scheme: ' .. uri)
```

Comment (h-michael) [reply runtime/lua/vim/uri.lua:79]: The scheme may have  also digit, "+", "-" and ".".
https://tools.ietf.org/html/rfc3986#page-17

**Review by tjdevries — APPROVED** [review]
LGTM to me as far as I can see.

**Review by h-michael — APPROVED** [review]

**Pull request merged**